### ISSUE TITLE
kernel: software CFI

### ISSUE THREAD

**Issue by samitolvanen** [google/android-riscv64#56]
Linux doesn't have KCFI support for RISC-V (`CONFIG_CFI_CLANG`). We should support this until #15 is available. Specifically, we need to implement KCFI lowering in the LLVM RISC-V back-end and error handling in the kernel. Most likely this is also going to involve fixing some function type mismatches in the kernel, but that's also going to be needed for hardware CFI.

**Comment by samitolvanen** [comment]
I tested this quickly with LLVM's generic KCFI implementation by simply applying this patch, and the ToT kernel booted just fine in qemu for me:
```diff
diff --git a/arch/riscv/Kconfig b/arch/riscv/Kconfig
index c5e42cc37604..e78936881899 100644
--- a/arch/riscv/Kconfig
+++ b/arch/riscv/Kconfig
@@ -40,6 +40,7 @@ config RISCV
        select ARCH_SUPPORTS_DEBUG_PAGEALLOC if MMU
        select ARCH_SUPPORTS_HUGETLBFS if MMU
        select ARCH_SUPPORTS_PAGE_TABLE_CHECK if MMU
+       select ARCH_SUPPORTS_CFI_CLANG
        select ARCH_USE_MEMTEST
        select ARCH_USE_QUEUED_RWLOCKS
        select ARCH_WANT_DEFAULT_TOPDOWN_MMAP_LAYOUT if MMU
```
LKDTM confirms that an indirect call with mismatching types indeed trap:
```
# echo CFI_FORWARD_PROTO > /sys/kernel/debug/provoke-crash/DIRECT 
[   18.190866] lkdtm: Performing direct entry CFI_FORWARD_PROTO     
[   18.191412] lkdtm: Calling matched prototype ...                                                                                         
[   18.191776] lkdtm: Calling mismatched prototype ...                                                                                      
[   18.192453] Oops - illegal instruction [#1]                        
[   18.192726] Modules linked in:                                     
[   18.192994] CPU: 0 PID: 102 Comm: sh Not tainted 6.3.0-rc1-00020-gfb7fb8611ef6-dirty #1
[   18.193422] Hardware name: riscv-virtio,qemu (DT)                  
[   18.193744] epc : lkdtm_indirect_call+0x2c/0x32                                                                                          
[   18.194373]  ra : lkdtm_CFI_FORWARD_PROTO+0x40/0x74                                                                                      
[   18.194645] epc : ffffffff8059de78 ra : ffffffff8059dd00 sp : ff200000004e3cf0                                                           
[   18.195039]  gp : ffffffff813095b0 tp : ff60000002411680 t0 : ff200000004e3b78
[   18.195515]  t1 : ff200000004e3ae8 t2 : 0000000000000018 s0 : ff200000004e3d00                                                           
[   18.196017]  s1 : 0000000000000001 a0 : 000000003ad55aca a1 : ffffffff8059de98                                                           
[   18.196634]  a2 : 000000007e0c52a5 a3 : 0000000000000001 a4 : 0000000000000fff
[   18.197176]  a5 : 0000000000000004 a6 : 00000000000000b7 a7 : 0000000000000000      
[   18.197718]  s2 : ff200000004e3e78 s3 : ffffffffffffffea s4 : 0000000000000012
[   18.198272]  s5 : ff60000002528000 s6 : 0000000000000006 s7 : ffffffff80e82e50                                                           
[   18.198963]  s8 : 0000000000000008 s9 : 0000000000000002 s10: ffffffff812bf258                                                           
[   18.199907]  s11: ffffffff812bf258 t3 : 0000000000000003 t4 : 0000000000000000
[   18.201205]  t5 : ffffffff8133df08 t6 : ffffffff8133df07           
[   18.202122] status: 0000000000000120 badaddr: 0000000000000000 cause: 0000000000000002
[   18.202927] [<ffffffff8059de78>] lkdtm_indirect_call+0x2c/0x32
[   18.203488] [<ffffffff8059dd00>] lkdtm_CFI_FORWARD_PROTO+0x40/0x74
[   18.204082] [<ffffffff8059a1cc>] lkdtm_do_action+0x20/0x34
[   18.204557] [<ffffffff8059a188>] direct_entry+0x124/0x136                                                                                
[   18.204880] [<ffffffff8033f60a>] full_proxy_write+0x56/0xb2                                                                              
[   18.205264] [<ffffffff801d5a36>] vfs_write+0x14a/0x338            
[   18.205527] [<ffffffff801d5d76>] ksys_write+0x64/0xd4                                                                                    
[   18.206024] [<ffffffff801d5df4>] sys_write+0xe/0x1a                                                                                      
[   18.206418] [<ffffffff800039a2>] ret_from_syscall+0x0/0x2                                                                                
[   18.207019] Code: 00c5 0517 00d7 0513 bea5 9582 60a2 6402 0141 8082 (0000) 52a5                                                          
[   18.208873] ---[ end trace 0000000000000000 ]---                                                                                         
[   18.209790] Kernel panic - not syncing: Fatal exception                                                                                  
[   18.210256] SMP: stopping secondary CPUs
```
Of course, there could be type mismatches hiding in code that wasn't exercised in this trivial test, and we might need to add type annotations to some assembly functions as well, but it looks like basic KCFI support should be fairly easy to implement.

**Comment by samitolvanen** [comment]
Here's an LLVM patch for RV64 KCFI operand bundle lowering: https://reviews.llvm.org/D148385

Tested in qemu with these kernel patches, which I'll send to LKML once we have some feedback on the compiler changes: https://github.com/samitolvanen/linux/commits/kcfi

**Comment by samitolvanen** [comment]
Recent upstream kernel changes exposed a function pointer type mismatch in syscall handling, which needs to be fixed before we can enable CFI in the kernel. Note that this is also going to be an issue with fine-grained Zisslpcfi. Details in https://github.com/ClangBuiltLinux/linux/issues/1843.

**Comment by samitolvanen** [comment]
LLVM commit: https://reviews.llvm.org/rG83835e22c7cd50c0e9b836cc359b6c59985b921f

v1 of the kernel patches: https://lore.kernel.org/lkml/20230629234244.1752366-8-samitolvanen@google.com/

**Comment by samitolvanen** [comment]
v2 of the kernel patches are "in the queue" but probably won't make it to Linux 6.5:

https://lore.kernel.org/linux-riscv/mhng-2fe7822f-f869-4513-b8a6-7da21ce5ccd8@palmer-ri-x1c9a/

**Comment by samitolvanen** [comment]
Merged to 6.6:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit?id=7f7d3ea6eb000bd329a6f2fe3f1c7596c4e783e1

Still needs to be enabled in android-mainline once the patches reach there.

**Comment by samitolvanen** [comment]
Enabled in android-mainline: https://r.android.com/2797165
